// @allmaps/render/dist/shared/events.js
var WarpedMapEventType = /* @__PURE__ */ ((WarpedMapEventType2) => {
  WarpedMapEventType2["GEOREFERENCEANNOTATIONADDED"] = "georeferenceannotationadded";
  WarpedMapEventType2["GEOREFERENCEANNOTATIONREMOVED"] = "georeferenceannotationremoved";
  WarpedMapEventType2["WARPEDMAPADDED"] = "warpedmapadded";
  WarpedMapEventType2["WARPEDMAPREMOVED"] = "warpedmapremoved";
  WarpedMapEventType2["WARPEDMAPENTER"] = "warpedmapenter";
  WarpedMapEventType2["WARPEDMAPLEAVE"] = "warpedmapleave";
  WarpedMapEventType2["IMAGEINFOLOADED"] = "imageinfoloaded";
  WarpedMapEventType2["TILEFETCHED"] = "tilefetched";
  WarpedMapEventType2["TILEFETCHERROR"] = "tilefetcherror";
  WarpedMapEventType2["MAPTILELOADED"] = "maptileloaded";
  WarpedMapEventType2["MAPTILEREMOVED"] = "maptileremoved";
  WarpedMapEventType2["FIRSTMAPTILELOADED"] = "firstmaptileloaded";
  WarpedMapEventType2["ALLREQUESTEDTILESLOADED"] = "allrequestedtilesloaded";
  WarpedMapEventType2["TEXTURESUPDATED"] = "texturesupdated";
  WarpedMapEventType2["PRECHANGE"] = "prechange";
  WarpedMapEventType2["OPTIONSCHANGED"] = "optionschanged";
  WarpedMapEventType2["ZINDICESCHANGED"] = "zindiceschanged";
  WarpedMapEventType2["RESOURCEMASKCHANGED"] = "resourcemaskchanged";
  WarpedMapEventType2["GCPSCHANGED"] = "gcpschanged";
  WarpedMapEventType2["VISIBILITYCHANGED"] = "visibilitychanged";
  WarpedMapEventType2["TRANSFORMATIONCHANGED"] = "transformationchanged";
  WarpedMapEventType2["DISTORTIONCHANGED"] = "distortionchanged";
  WarpedMapEventType2["INTERNALPROJECTIONCHANGED"] = "internalprojectionchanged";
  WarpedMapEventType2["PROJECTIONCHANGED"] = "projectionchanged";
  WarpedMapEventType2["CHANGED"] = "changed";
  WarpedMapEventType2["CLEARED"] = "cleared";
  return WarpedMapEventType2;
})(WarpedMapEventType || {});
class WarpedMapEvent extends Event {
  data;
  constructor(type, data) {
    super(type);
    this.data = data;
  }
}
export {
  WarpedMapEvent,
  WarpedMapEventType
};
