// @allmaps/transform/dist/shared/conversion-functions.js
// Invert
export function invertGeneralGcp(generalGcp) {
    return { source: generalGcp.destination, destination: generalGcp.source };
}
export function invertGcp(gcp) {
    return { resource: gcp.geo, geo: gcp.resource };
}
// Select
export function generalGcpToPointForForward(generalGcp) {
    return generalGcp.destination;
}
export function generalGcpToPointForBackward(generalGcp) {
    return generalGcp.source;
}
export function gcpToPointForToGeo(gcp) {
    return gcp.geo;
}
export function gcpToPointForToResource(gcp) {
    return gcp.resource;
}
// GeneralGcp <> G
export function generalGcpToGcp(generalGcp) {
    return {
        resource: generalGcp.source,
        geo: generalGcp.destination,
        partialDerivativeX: generalGcp.partialDerivativeX,
        partialDerivativeY: generalGcp.partialDerivativeY,
        distortions: generalGcp.distortions,
        distortion: generalGcp.distortion
    };
}
export function gcpToGeneralGcp(gcp) {
    return {
        source: gcp.resource,
        destination: gcp.geo,
        partialDerivativeX: gcp.partialDerivativeX,
        partialDerivativeY: gcp.partialDerivativeY,
        distortions: gcp.distortions,
        distortion: gcp.distortion
    };
}
